cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,426 @@
+   * @brief Get a device array containing the depth of each column in the preprocessed table
+   *
+   * @see struct_linearize()
+   *
+   * @return std::optional<device_span<int const>> Device array containing respective column depths.
+   * If there are no nested columns in the table then this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<int const>> depths() const
+  {
+    return d_depths.size() ? std::optional<device_span<int const>>(d_depths) : std::nullopt;
+  }
+
+  /**
+   * @brief Whether the table has any nullable column
+   *
+   */
+  [[nodiscard]] bool has_nulls() const { return _has_nulls; }
+
+ private:
+  using table_device_view_owner =
+    std::invoke_result_t<decltype(table_device_view::create), table_view, rmm::cuda_stream_view>;
+
+  std::unique_ptr<table_device_view_owner> d_t;
+  rmm::device_uvector<order> d_column_order;
+  rmm::device_uvector<null_order> d_null_precedence;
+  rmm::device_uvector<size_type> d_depths;
+  bool _has_nulls;
+};
+
+/**
+ * @brief An owning object that can be used to lexicographically compare two rows of the same table
+ *
+ * This class can take a table_view and preprocess certain columns to allow for lexicographical
+ * comparison. The preprocessed table and temporary data required for the comparison are created and
+ * owned by this class.
+ *
+ * This class can then provide a functor object that can used on the device.
+ * The object of this class must outlive the usage of the device functor.
+ */
+class self_comparator {
+ public:
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same table.
+   *
+   * @param table The table to compare
+   * @param column_order Optional, host array the same length as a row that indicates the desired
+   * ascending/descending order of each column in a row. If empty, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If empty, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   * @param stream The stream to construct this object on. Not the stream that will be used for
+   * comparisons using this object.
+   */
+  self_comparator(table_view const& t,
+                  host_span<order const> column_order         = {},
+                  host_span<null_order const> null_precedence = {},
+                  rmm::cuda_stream_view stream                = rmm::cuda_stream_default)
+    : d_t{std::make_shared<preprocessed_table>(t, column_order, null_precedence, stream)}
+  {
+  }
+
+  /**
+   * @brief Construct an owning object for performing a lexicographic comparison between two rows of
+   * the same preprocessed table.
+   *
+   * @param t A table preprocessed for lexicographic comparison
+   */
+  self_comparator(std::shared_ptr<preprocessed_table> t) : d_t{std::move(t)} {}
```

Comment:
Does this constructor ever need to be exposed, or can it be private? If it can be private, I guess that leads me to two additional question:
1) Does it ever need to be used even internally? I don't see anywhere, but I could be missing it.
2) If this is never used externally, do we even need to expose the `preprocessed_table`, or can we insulate users of this functor from the internal detail of preprocessing entirely?

Reply:
Good question. The previous design of owning `self_comparator` hid all the preprocessing details from the user. But an issue came up: there are places where a single `table_device_view` is used in multiple comparators (e.g. in join). The old non-owning comparator is cheap to construct and copy because the expensive part is creating the `table_device_view`. Now the same use case can be supported by constructing a `preprocessed_table` and passing it to multiple owning comparators, while users who don't need to do such stuff have a convenient method without worrying about preprocessing (see sort_impl changes)